### PR TITLE
Avoid AmbigMember in lookup of interfaces with nullability differences only

### PR DESCRIPTION
**Pull request by jcouv**
Scenarios where an interface is implemented twice with nullability differences result in lookup _errors_.
See example below. My PR on type substitution (https://github.com/dotnet/roslyn/pull/48694) would make this more common (wouldn't require the `class` constraint) and ran into this error during bootstrap build.

The general desire for the nullability feature is that it should only produce warnings.

Options:
1. merge types and their nullabilities (see `TypeWithAnnotations.MergeEquivalentTypes`), so that `AllInterfaces` wouldn't contain such duplicates
2. merges types and their nullabilities for purpose of lookup (see `GetBaseInterfaces` in `Binder_Lookup.cs`)
3. pick a winner whenever such conflicts are detected during lookup

From discussion with Aleksey, option 3 seems the safest, as it does not create types that didn't exist. It's possible that parts of the code base assume that `.Interfaces` is a subset of `.AllInterfaces`, for instance, or that members found during lookup on an interface belong to that interface or its bases.

Open question: should we gate this fix behing LangVer in some way?

```csharp
#nullable disable
interface I<T> where T : class { T Item { get; } }

#nullable enable
interface I2<T> : I<T> where T : class { }

#nullable disable
interface I3 : I<object>, I2<object> { }

#nullable enable
public class C
{
    void M(I3 i)
    {
        _ = i.Item; // error CS0229: Ambiguity between 'I<object>.Item' and 'I<object>.Item'
    }
}
```

### CODE REVIEW

**Comment by RikkiGibson** [comment]
> Open question: should we gate this fix behing LangVer in some way?

At a glance it feels like no to me. This is a bug-fix level change, especially since in old compilers use of the `class` constraint will also be needed to get a unification failure.

**Comment by jaredpar** [comment]
Agree. Don't gate on `langver`

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1151]: For my eductation: can we get into a similar problem when interfaces differ by `object` and `dynamic` type parameters?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1151]: yes, but `dynamic` is not allowed as a type argument. See `AmbigMember_DynamicDifferences` below.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1153]: >goto symIsHidden [](start = 32, length = 16)

How do we know it is the "same" member? We are not looking at signature etc. For example, in the ```if``` below that is also deals with "hiding" we don't make this assumption. I am assuming that we are "resolving" that "hiding" elsewhere. It feels like the new hiding should be resolved there as well. Meaning that we simply add negated ```if``` condition to the ```if``` below and add additional code elsewhere to handle methods and indexers.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142092]: >comp.VerifyDiagnostics(); [](start = 12, length = 25)

In this and other tests that are supposed to verify that an ambiguity is successfully resolved, I think we should verify how exactly it is resolved to ensure that the resolution is deterministic, etc.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142087]: >Item [](start = 10, length = 4)

Consider expanding testing to other kinds of members: fields, indexers, events, nested types, methods.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1153]: It looks like alternatively we could filter out "duplicate" interfaces inside ```GetBaseInterfaces```` helper.

---
In reply to: [524690053](https://github.com/dotnet/roslyn/pull/49338#discussion_r524690053) [](ancestors = 524690053)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142092]: We're already verifying this in two ways: the behavior of `.Item` during nullability analysis, and by checking how the ambiguity is resolved in `AllInterfaces`.
I could also check `GetSymbolInfo` on `i.Item`, but that doesn't seem to add much. Is that what you had in mind?

---
In reply to: [524699086](https://github.com/dotnet/roslyn/pull/49338#discussion_r524699086) [](ancestors = 524699086)

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1153]: I don't think we can filter in `GetBaseInterfaces` because that would affect `IsDerivedType` which also uses that helper, but we could do the filtering at the call-site of `GetBaseInterfaces` in `LookupMembersInInterfaceOnly`.
I had done that in drafting this PR, but changed to this approach to reduce allocations and iterations.
Do you have a preference?

I didn't fully understand your question above. We can discuss tomorrow.

---
In reply to: [524709873](https://github.com/dotnet/roslyn/pull/49338#discussion_r524709873) [](ancestors = 524709873,524690053)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142092]: > and by checking how the ambiguity is resolved in ```AllInterfaces```

As far as I can tell, there is no change to the behavior of this API. What could we be checking?

> I could also check GetSymbolInfo on i.Item, but that doesn't seem to add much. Is that what you had in mind?

Yes. something like that, I think we need to assert specific symbol. We probably should also test lookup APIs that SemanticModel offers.

---
In reply to: [542852918](https://github.com/dotnet/roslyn/pull/49338#discussion_r542852918) [](ancestors = 542852918,524699086)

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1153]: > I don't think we can filter in GetBaseInterfaces because that would affect IsDerivedType which also uses that helper

That is a fair point. Then it looks like we could filter interfaces in ```LookupMembersInInterfacesWithoutInheritance```. Why bother even performing a lookup if we already looked in a "similar" interface and planning to discard everything we find?

> I didn't fully understand your question above. 

C# hides by signature. Signature is part of detecting if the candidate is different (as the comment says). Where do we check the signature if we are saying "as it is not really different"? If we don't check the signature, we shouldn't claim that "it is not really different". 

---
In reply to: [542952838](https://github.com/dotnet/roslyn/pull/49338#discussion_r542952838) [](ancestors = 542952838,524709873,524690053)

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1061]: >HashSet [](start = 16, length = 7)

Would it make sense to use pooled hash set?

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1069]: >seenInterfaces is null || !seenInterfaces.Contains(baseInterface) [](start = 24, length = 65)

```seenInterfaces?.Add(baseInterface) != false``` and get rid of the ```Add``` inside the ```if```?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142306]: >.Contains("Item") [](start = 79, length = 17)

I think we actually want to assert in a way that would allow us to verify from which flavor of the interface this symbol actually came from. We want to see is there an ambiguity here and how it is resolved.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142339]: >ToTestDisplayString [](start = 127, length = 19)

Should we use ```TypeWithAnnotations.TestDisplayFormat``` here?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142437]: >Contains("Get") [](start = 80, length = 15)

Similar comment as for the other test above.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:141943]: >void M(I3 i) [](start = 4, length = 12)

I think we should also test lookup in a type parameter. I.e. when ```i``` is of type ```T```, which has ```I3``` as a constraint.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142103]: >ToTestDisplayString [](start = 111, length = 19)

TypeWithAnnotations.TestDisplayFormat

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142283]: >I<object>, I2<object> [](start = 15, length = 21)

Does flipping the order here change what symbols we get?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:141943]: > I think we should also test lookup in a type parameter. I.e. when i is of type T, which has I3 as a constraint.

This comment is for scenarios involving the nullability mismatch. Also, I think we should test scenarios when  constraints are just ```I<object>, I2<object>``` and ```I2<object>, I<object>```.

---
In reply to: [544360179](https://github.com/dotnet/roslyn/pull/49338#discussion_r544360179) [](ancestors = 544360179)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142283]: It does. See `AmbigMember_DifferenceBetweenNonnullableAndOblivious_WithConstraint_ReverseOrder`

---
In reply to: [544374723](https://github.com/dotnet/roslyn/pull/49338#discussion_r544374723) [](ancestors = 544374723)

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:1061]: The reason I didn't use the pool is that we'd need a new one for this specific comparer. Not sure that's warranted

---
In reply to: [544338611](https://github.com/dotnet/roslyn/pull/49338#discussion_r544338611) [](ancestors = 544338611)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:142283]: >It does. See AmbigMember_DifferenceBetweenNonnullableAndOblivious_WithConstraint_ReverseOrder

It doesn't look like we assert symbol in that test.

---
In reply to: [544603988](https://github.com/dotnet/roslyn/pull/49338#discussion_r544603988) [](ancestors = 544603988,544374723)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5), assuming CI is passing.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

File: src/Compilers/Test/Utilities/CSharp/SymbolUtilities.cs
Comment (RikkiGibson) [inline src/Compilers/Test/Utilities/CSharp/SymbolUtilities.cs:125]: nit: consider doing something like the following here:

```cs
            return symbols.SelectAsArray(static (s, format) => s.ToDisplayString(format), format);
```

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/Test/Utilities/CSharp/SymbolUtilities.cs
Comment (jcouv) [reply src/Compilers/Test/Utilities/CSharp/SymbolUtilities.cs:125]: I'll leave as-is to get this into preview3. Thanks

**Pull request merged**
